codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsJsonRpc1_0.java

```diff
@@ -36,4 +45,73 @@ protected String getDocumentContentType() {
+                context, operation, responseType, this::writeErrorMessageCodeDeserializer,
+                this::getOperationErrors, (writer) -> AwsJsonRpc1_0.awsQueryCompatibleDefaultBlockWriter(writer));
+        } else {
+            return HttpProtocolGeneratorUtils.generateErrorDispatcher(
+                context, operation, responseType, this::writeErrorMessageCodeDeserializer,
+                this::getOperationErrors);
+        }
+    }
+
+    @Override
+    protected void deserializeError(GenerationContext context, StructureShape shape) {
+        GoWriter writer = context.getWriter().get();
+        Symbol symbol = context.getSymbolProvider().toSymbol(shape);
+        String functionName = ProtocolGenerator.getDocumentDeserializerFunctionName(shape, context.getService(),
+                getProtocolName());
+
+        AwsProtocolUtils.initializeJsonDecoder(writer, "errorBody");
+        AwsProtocolUtils.decodeJsonIntoInterface(writer, "");
+        writer.write("output := &$T{}", symbol);
+        writer.write("err := $L(&output, shape)", functionName);
+        writer.write("");
+        AwsProtocolUtils.handleDecodeError(writer);
+        writer.write("errorBody.Seek(0, io.SeekStart)");
+        if (context.getService().hasTrait(AwsQueryCompatibleTrait.class)) {
+            writer.write("codeFromHeader := getAwsQueryErrorCode(response)");
+            writer.openBlock("if codeFromHeader != \"\" {", "}", () -> {
+                writer.write("output.ErrorCodeOverride = &codeFromHeader");
+            });
+        }
+        writer.write("return output");
+    }
+
+    @Override
+    public void generateSharedDeserializerComponents(GenerationContext context) {
+        super.generateSharedDeserializerComponents(context);
+        if (isAwsQueryCompatibleTraitFound(context)) {
+            GoWriter writer = context.getWriter().get();
+            writer.openBlock("func getAwsQueryErrorCode(response *smithyhttp.Response) string {", "}", () -> {
+                writer.write("queryCodeHeader := response.Header.Get(\"x-amzn-query-error\")");
+                writer.openBlock("if queryCodeHeader != \"\" {", "}", () -> {
+                    writer.write("queryCodeParts := strings.Split(queryCodeHeader, \";\")");
+                    writer.openBlock("if queryCodeParts != nil && len(queryCodeParts) == 2 {", "}", () -> {
+                        writer.write("return queryCodeParts[0]");
+                    });
+                });
+                writer.write("return \"\"");
+            });
+        }
+    }
+
+    private static void awsQueryCompatibleDefaultBlockWriter(GoWriter writer) {
+        writer.openBlock("default:", "", () -> {
+            writer.openBlock("genericError := &smithy.GenericAPIError{", "}", () -> {
+                writer.write("Code: getAwsQueryErrorCode(response),");
+                writer.write("Message: errorMessage,");
+            });
+            writer.write("return genericError");
```

Comment:
Add one more check to keep `errorCode`'s original value in case `getAwsQueryErrorCode()` returns `""`:

```suggestion
        writer.openBlock("default:", "", () -> {
            writer.write("awsQueryErrorCode := getAwsQueryErrorCode(response)");
            writer.openBlock("if awsQueryErrorCode != \"\" {", "}", () -> {
                writer.write("errorCode = awsQueryErrorCode");
            });
            writer.openBlock("genericError := &smithy.GenericAPIError{", "}", () -> {
                writer.write("Code: errorCode,");
                writer.write("Message: errorMessage,");
            });
            writer.write("return genericError");
```

Diff:

```go
    default:
        genericError := &smithy.GenericAPIError{
            Code:    getAwsQueryErrorCode(response),
            Message: errorMessage,
        }
        return genericError
```

to 

```go
    default:
        awsQueryErrorCode := getAwsQueryErrorCode(response)
        if awsQueryErrorCode != "" {
            errorCode = awsQueryErrorCode
        }
        genericError := &smithy.GenericAPIError{
            Code:    errorCode
            Message: errorMessage,
        }
        return genericError
```